Enhanced <select>: element nesting changes

annevk opened the pull request:
#### 71fdfc0b942fc5eb8547d2963c0325dd6db46d08
<pre>
Enhanced &lt;select&gt;: element nesting changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=300317">https://bugs.webkit.org/show_bug.cgi?id=300317</a>
<a href="https://rdar.apple.com/162596003">rdar://162596003</a>

Reviewed by Ryosuke Niwa.

&lt;select&gt;, &lt;optgroup&gt;, &lt;option&gt;, and &lt;hr&gt; elements can now be more
arbitrarily nested.

We also ensure that &lt;select&gt;.options and &lt;select&gt;.selectedOptions
only return &lt;option&gt; elements that belong to the &lt;select&gt; element in
question.

All of this is gated behind the HTMLEnhancedSelectParsingEnabled
preference as this new behavior is only really relevant if the HTML
parser supports such arbitrary nesting.

Tests:

- <a href="https://github.com/web-platform-tests/wpt/pull/55419">https://github.com/web-platform-tests/wpt/pull/55419</a>
- <a href="https://github.com/web-platform-tests/wpt/pull/55424">https://github.com/web-platform-tests/wpt/pull/55424</a>

This does not handle dynamic changes for &lt;option&gt; element descendants.

Canonical link: <a href="https://commits.webkit.org/302288@main">https://commits.webkit.org/302288@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7de9821cfa01b5423a74ae9f0c8ce32e84bf3a49

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128445 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/715 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39276 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135840 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79893 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/57d73f78-c504-4b3f-93ae-359d32cad6aa) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130317 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/658 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/588 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97778 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65681 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/75671271-5f5e-4a56-9df0-be1e8d2cf3fa) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131393 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/472 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115073 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78389 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8b76b361-9aab-4507-bf98-efe52341dde2) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/433 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33180 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79122 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108839 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33664 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138289 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/524 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106319 "Found 1 new test failure: imported/blink/fast/pagination/viewport-y-horizontal-bt-rtl.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/596 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111414 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106129 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27066 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/459 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29957 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52881 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/607 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63800 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/501 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/561 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/565 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -56,6 +56,28 @@ Ref<HTMLOptGroupElement> HTMLOptGroupElement::create(const QualifiedName& tagNam
+            select->setRecalcListItems();
+    }
+
+    HTMLElement::removedFromAncestor(removalType, oldParentOfRemovedTree);
```
We need to call this function before doing the traversal. Otherwise, Node flags are not yet updated.

annevk replied:
Note that we do it the other way around in `HTMLArticleElement::removedFromAncestor` at least. And Chromium also generally seems to do it after. I'm not familiar enough with this code though so I'll make those changes.

rniwa replied:
It's generally a mistake to do any work before calling `Node::removedFromAncestor`.

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -131,7 +153,16 @@ String HTMLOptGroupElement::groupLabelText() const
+    if (!document().settings().htmlEnhancedSelectParsingEnabled())
+        return dynamicDowncast<HTMLSelectElement>(parentNode());
+
+    for (Ref ancestor : ancestorsOfType<HTMLElement>(*const_cast<HTMLOptGroupElement*>(this))) {
```
We should start the search from `parentNode()`. Then you don't need a `const_cast` here.

annevk replied:
Unfortunately that doesn't work as it's ancestorsOfType and not inclusiveAncestorsOfType, so we'd end up skipping a node. Maybe we should introduce the latter though so we can avoid const_cast here and there? Probably best as a separate change.

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -131,7 +153,16 @@ String HTMLOptGroupElement::groupLabelText() const
+        return dynamicDowncast<HTMLSelectElement>(parentNode());
+
+    for (Ref ancestor : ancestorsOfType<HTMLElement>(*const_cast<HTMLOptGroupElement*>(this))) {
+        if (RefPtr select = dynamicDowncast<HTMLSelectElement>(ancestor))
```
Since there is no non-trivial code here, you can just use a raw pointer (`auto *`).

rniwa commented on Source/WebCore/html/HTMLOptionElement.cpp:
```diff
@@ -90,6 +91,28 @@ ExceptionOr<Ref<HTMLOptionElement>> HTMLOptionElement::createForLegacyFactoryFun
+            select->setRecalcListItems();
+    }
+
+    HTMLElement::removedFromAncestor(removalType, oldParentOfRemovedTree);
```
Need to call this one first.

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -131,7 +153,16 @@ String HTMLOptGroupElement::groupLabelText() const
+    if (!document().settings().htmlEnhancedSelectParsingEnabled())
+        return dynamicDowncast<HTMLSelectElement>(parentNode());
+
+    for (Ref ancestor : ancestorsOfType<HTMLElement>(*const_cast<HTMLOptGroupElement*>(this))) {
```
Why we can't just call `HTMLSelectElement::findAncestorSelectElement` here?

rniwa commented on Source/WebCore/html/HTMLHRElement.cpp:
```diff
@@ -55,6 +55,28 @@ Ref<HTMLHRElement> HTMLHRElement::create(const QualifiedName& tagName, Document&
+            select->setRecalcListItems();
+    }
+
+    HTMLElement::removedFromAncestor(removalType, oldParentOfRemovedTree);
```
Need to call this one first. Otherwise Node flags are not yet updated.

rniwa commented on Source/WebCore/html/HTMLOptionElement.cpp:
```diff
@@ -269,13 +292,16 @@ void HTMLOptionElement::childrenChanged(const ChildChange& change)
-            return optGroup->ownerSelectElement();
+    if (!document().settings().htmlEnhancedSelectParsingEnabled()) {
+        if (RefPtr parent = parentElement()) {
+            if (RefPtr select = dynamicDowncast<HTMLSelectElement>(*parent))
```
No need to use RefPtr here since all the code in the scope is trivial.

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -101,6 +101,19 @@ Ref<HTMLSelectElement> HTMLSelectElement::create(Document& document)
+HTMLSelectElement* HTMLSelectElement::findAncestorSelectElement(ContainerNode& node)
+{
+    for (Ref ancestor : ancestorsOfType<Element>(node)) {
+        if (RefPtr select = dynamicDowncast<HTMLSelectElement>(ancestor))
```
No need to use `RefPtr` here. Just use `auto *`.

rniwa commented on Source/WebCore/html/HTMLHRElement.cpp:
```diff
@@ -55,6 +55,36 @@ Ref<HTMLHRElement> HTMLHRElement::create(const QualifiedName& tagName, Document&
+{
+    auto result = HTMLElement::insertedIntoAncestor(insertionType, parentOfInsertedTree);
+
+    if (document().settings().htmlEnhancedSelectParsingEnabled()) {
```
I think it's better do an early return here just like `removedFromAncestor`.

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -56,6 +56,37 @@ Ref<HTMLOptGroupElement> HTMLOptGroupElement::create(const QualifiedName& tagNam
+    if (!document().settings().htmlEnhancedSelectParsingEnabled())
+        return result;
+
+    if (auto maybeSelect = HTMLSelectElement::findAncestorSelectElement(*this, HTMLSelectElement::ExcludeOptGroup::Yes, &parentOfInsertedTree)) {
```
I think we should store this result using WeakPtr so that we don't have to do two tree traversal to look for `select` element upon removal.

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -101,6 +102,30 @@ Ref<HTMLSelectElement> HTMLSelectElement::create(Document& document)
     return adoptRef(*new HTMLSelectElement(selectTag, document, nullptr));
 }
 
+std::optional<HTMLSelectElement*> HTMLSelectElement::findAncestorSelectElement(ContainerNode& node, ExcludeOptGroup excludeOptGroup, ContainerNode* mustSeeNode)
```
This is a very convoluted type to return since the semantic difference between `std::nullopt` and `nullptr` isn't self evident at glance.
I think we should change this to return `Expected` with an appropriate enum class values.

annevk replied:
With `m_ownerSelect` this could be simplified to `HTMLSelectElement*`.

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -56,6 +56,37 @@ Ref<HTMLOptGroupElement> HTMLOptGroupElement::create(const QualifiedName& tagNam
+
+    if (auto maybeSelect = HTMLSelectElement::findAncestorSelectElement(*this, HTMLSelectElement::ExcludeOptGroup::Yes, &parentOfInsertedTree)) {
+        if (RefPtr select = *maybeSelect)
+            select->setRecalcListItems();
```
Do we need to do this even when this element isn't connected to a document?
If this is purely for style recalc, we most certainly don't need to all this work if the element is still disconnected.

annevk replied:
We need this for `HTMLSelectElement::listItems()` to be correct. It essentially has a cache of the underlying list and the API assumes that is always in place, even when it's disconnected.

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -56,6 +56,37 @@ Ref<HTMLOptGroupElement> HTMLOptGroupElement::create(const QualifiedName& tagNam
+
+    if (auto maybeSelect = HTMLSelectElement::findAncestorSelectElement(oldParentOfRemovedTree, HTMLSelectElement::ExcludeOptGroup::Yes)) {
+        if (RefPtr select = *maybeSelect)
+            select->setRecalcListItems();
```
Ditto. Do we need to do this work when the tree from which this subtree was removed was disconnected to begin with?

rniwa commented on Source/WebCore/html/HTMLHRElement.cpp:
```diff
@@ -55,6 +55,37 @@ Ref<HTMLHRElement> HTMLHRElement::create(const QualifiedName& tagName, Document&
+    if (!document().settings().htmlEnhancedSelectParsingEnabled() || !m_ownerSelect)
+        return;
+
+    if (RefPtr select = HTMLSelectElement::findOwnerSelect(protectedParentNode().get(), HTMLSelectElement::ExcludeOptGroup::Yes)) {
```
We should wrap this if-statement in `#if ASSERT_ENABLED` so that we don't do the tree traversal in release builds.

annevk replied:
The early return is needed to avoid updating select in this case. It's not just an assert.

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -56,6 +56,37 @@ Ref<HTMLOptGroupElement> HTMLOptGroupElement::create(const QualifiedName& tagNam
+    if (!document().settings().htmlEnhancedSelectParsingEnabled() || !m_ownerSelect)
+        return;
+
+    if (RefPtr select = HTMLSelectElement::findOwnerSelect(protectedParentNode().get(), HTMLSelectElement::ExcludeOptGroup::Yes)) {
```
Ditto. We should avoid the traversal in release builds.

rniwa commented on Source/WebCore/html/HTMLHRElement.cpp:
```diff
@@ -55,6 +55,37 @@ Ref<HTMLHRElement> HTMLHRElement::create(const QualifiedName& tagName, Document&
+        return;
+    }
+
+    if (RefPtr select = std::exchange(m_ownerSelect, nullptr).get())
```
This isn't quite right. I don't think we should be clearing m_ownerSelect when m_ownerSelect got removed along with this `hr` element. Or is `setRecalcListItems` only used when `select` is in a document? If so, we should check `removalType.disconnectedFromDocument`.

annevk replied:
That's why we return early above. So we only update if we got removed from the select. But that can also happen if the select is not in the document so I don't think we can do what you suggest.

rniwa commented on Source/WebCore/html/HTMLOptGroupElement.cpp:
```diff
@@ -56,6 +56,37 @@ Ref<HTMLOptGroupElement> HTMLOptGroupElement::create(const QualifiedName& tagNam
+        return;
+    }
+
+    if (RefPtr select = std::exchange(m_ownerSelect, nullptr).get())
```
Ditto. This isn't quite right.

rniwa commented on Source/WebCore/html/HTMLOptionElement.cpp:
```diff
@@ -90,6 +91,37 @@ ExceptionOr<Ref<HTMLOptionElement>> HTMLOptionElement::createForLegacyFactoryFun
+    if (!document().settings().htmlEnhancedSelectParsingEnabled() || !m_ownerSelect)
+        return;
+
+    if (RefPtr select = HTMLSelectElement::findOwnerSelect(protectedParentNode().get(), HTMLSelectElement::ExcludeOptGroup::No)) {
```
Ditto.

rniwa commented on Source/WebCore/html/HTMLOptionElement.cpp:
```diff
@@ -90,6 +91,37 @@ ExceptionOr<Ref<HTMLOptionElement>> HTMLOptionElement::createForLegacyFactoryFun
+        return;
+    }
+
+    if (RefPtr select = std::exchange(m_ownerSelect, nullptr).get())
```
Ditto.

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -842,15 +863,57 @@ void HTMLSelectElement::recalcListItems(bool updateSelectedStates, AllowStyleInv
-            handleOptionElement(*option);
-        else if (is<HTMLHRElement>(child.get()))
-            m_listItems.append(&child.get());
+    if (document().settings().htmlEnhancedSelectParsingEnabled()) {
```
`childrenChanged` is called for each child we're inserting. We want to do this traversal lazily when someone asks for the updated list.

annevk replied:
This is done lazily as far as I know. This is a modification of recalcListItems, not childrenChanged.

annevk committed:
Enhanced <select>: element nesting changes
https://bugs.webkit.org/show_bug.cgi?id=300317
rdar://162596003

Reviewed by Ryosuke Niwa.

<select>, <optgroup>, <option>, and <hr> elements can now be more
arbitrarily nested.

We also ensure that <select>.options and <select>.selectedOptions
only return <option> elements that belong to the <select> element in
question.

All of this is gated behind the HTMLEnhancedSelectParsingEnabled
preference as this new behavior is only really relevant if the HTML
parser supports such arbitrary nesting.

Tests:

- https://github.com/web-platform-tests/wpt/pull/55419
- https://github.com/web-platform-tests/wpt/pull/55424

This does not handle dynamic changes for <option> element descendants.

Canonical link: https://commits.webkit.org/302288@main
